src/Razor/src/Microsoft.AspNetCore.Razor.Language/DefaultBoundAttributeDescriptorBuilder.cs

```diff
@@ -182,14 +183,29 @@ private IEnumerable<RazorDiagnostic> Validate()
                     yield return diagnostic;
                 }
 
-                foreach (var character in Name)
+                var name = Name;
+                if (isDirectiveAttribute && name.StartsWith("@"))
+                {
+                    name = name.Substring(1);
+                }
+                else if (isDirectiveAttribute)
+                {
+                    var diagnostic = RazorDiagnosticFactory.CreateTagHelper_InvalidBoundDirectiveAttributeName(
+                            _parent.GetDisplayName(),
+                            GetDisplayName(),
+                            Name);
+
+                    yield return diagnostic;
+                }
+
+                foreach (var character in name)
                 {
                     if (char.IsWhiteSpace(character) || HtmlConventions.InvalidNonWhitespaceHtmlCharacters.Contains(character))
                     {
                         var diagnostic = RazorDiagnosticFactory.CreateTagHelper_InvalidBoundAttributeName(
                             _parent.GetDisplayName(),
                             GetDisplayName(),
-                            Name,
+                            name,
```

Comment:
I did that on purpose because of the error message. Imagine your directive attribute looks like `@bi@nd`. It says something like, `This attribute is invalid because "@bi@nd" contains the "@" character`. This is a bit misleading. I don't feel strongly though.

Reply:
I see. I think it might be more misleading to have users searching for where they typed `bi@nd` and not realize that we removed the `@`.  I'd recommend changing the error message to say `contains the "@" character in its name` or something else a little more specific to aid in the identification of what went wrong.